File: csharp/EviFluor/storage.cs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;
namespace Hse.EviFluor;

/// <summary>
/// Represents an entry in the storage measurement system.
/// Contains a measurement, an optional comment, results, and raw JSON data.
/// </summary>
public class StorageMeasurementEntry
{
    /// <summary>
    /// The stored measurement data.
    /// </summary>
    public Measurement Measurement { get; set; }

    /// <summary>
    /// An optional comment associated with the measurement entry.
    /// </summary>
    public String? Comment { get; set; }

    /// <summary>
    /// The calculated results associated with the measurement, if available.
    /// </summary>
    public Results? Results { get; set; }

    /// <summary>
    /// The raw JSON node representation of the measurement entry.
    /// </summary>
    public JsonNode? Node { get; set; }

    /// <summary>
    /// Initializes a new instance of the <see cref="StorageMeasurementEntry"/> class.
    /// </summary>
    /// <param name="measurement">The measurement data.</param>
    /// <param name="comment">An optional comment.</param>
    /// <param name="results">The associated results, if available.</param>
    /// <param name="node">The raw JSON node, if available.</param>
    public StorageMeasurementEntry(Measurement measurement, String? comment = null, Results? results = null, JsonNode? node = null)
    {
        Measurement = measurement;
        Comment = comment;
        Results = results;
        Node = node;
    }

    /// <summary>
    /// Returns a string representation of the storage measurement entry.
    /// </summary>
    /// <returns>A formatted string displaying measurement details.</returns>
    public override string ToString()
    {
        return $"Measurement:{Measurement} Comment:{Comment} Results:{Results}";
    }

    /// <summary>
    /// Determines whether the entry contains calculated results.
    /// </summary>
    /// <returns><c>true</c> if results exist, otherwise <c>false</c>.</returns>
    public bool HasResults()
    {
        return Results != null;
    }

    /// <summary>
    /// Applies the calculated results based on given correction factors.
    /// Updates the JSON node representation if available.
    /// </summary>
    /// <param name="factors">The correction factors to apply.</param>
    /// <param name="kit">Optional kit used to calculate concentrations from the corrected values.</param>
    /// <remarks>
    /// When the entry was loaded from JSON (<see cref="Node"/> != <c>null</c>), the computed results are
    /// written back into that JSON node under <c>"results"</c>.
    /// </remarks> 
    public void ApplyResults(Factors factors, IKit? kit = null)
    {
        Results = Measurement.GetResults(factors, kit);
        if (Node != null)
        {
            Node[Dict.RESULTS] = Results.ToJson();
        }
    }

    /// <summary>
    /// Creates a <see cref="StorageMeasurementEntry"/> instance from a JSON node.
    /// Parses the measurement, comment, and results data.
    /// </summary>
    /// <param name="node">The JSON node containing the measurement entry data.</param>
    /// <returns>A populated <see cref="StorageMeasurementEntry"/> instance.</returns>
    public static StorageMeasurementEntry FromJson(JsonNode node)
    {
        Results? results = null;
        String? comment = null;

        if (node.AsObject().ContainsKey(Dict.RESULTS))
        {
            results = Results.FromJson(node[Dict.RESULTS]);
        }

        if (node.AsObject().ContainsKey(Dict.COMMENT))
        {
            if (node is JsonObject jsonObject &&
            jsonObject.TryGetPropertyValue(Dict.COMMENT, out JsonNode? commentNode) &&
            commentNode is not null)
            {
                comment = commentNode.GetValue<string>();
            }
        }
        return new StorageMeasurementEntry(Measurement.FromJson(node), comment, results, node);
    }
}


/// <summary>
/// Handles storage and retrieval of measurements using JSON serialization.
/// </summary>
public class StorageMeasurement
{
    /// <summary>
    /// Raw JSON structure representing all stored measurement data.
    /// This node may contain measurements, results, metadata, logging, and other associated entries.
    /// </summary>
    public JsonNode data { get; }


    /// <summary>
    /// Initializes a new instance of the <see cref="StorageMeasurement"/> class, optionally loading from a file.
    /// </summary>
    /// <param name="filename">Optional filename to load measurements from.</param>
    public StorageMeasurement(string filename = "")
    {
        if (string.IsNullOrEmpty(filename))
        {
            data = new JsonObject();
            data.AsObject()[Dict.MEASUREMENTS] = new JsonArray();
        }
        else
        {
            using (var reader = new StreamReader(filename))
            {
                var json = reader.ReadToEnd();
                data = JsonNode.Parse(json) ?? new JsonObject();
            }
        }
    }

    /// <summary>
    /// Appends a measurement to storage.
    /// </summary>
    /// <param name="measurement">The measurement to append.</param>
    /// <param name="comment">An optional comment for the measurement.</param>
    /// <param name="logging">Optional logging information.</param>
    /// <param name="verification">Optional verification information.</param>
    /// <remarks>
    /// Adds <c>date_time</c> in ISO-8601 UTC and optional <c>logging</c>/<c>errors</c> sections if provided.
    /// </remarks>
    public void Append(Measurement measurement, string comment = "", List<string> ? logging = null, Verification? verification = null)
    {
        if (measurement == null)
            throw new ArgumentException("No measurement object provided to append!");

        var m = measurement.ToJson();
        if (!string.IsNullOrEmpty(comment))
        {
            m[Dict.COMMENT] = comment;
        }

        if (logging != null && logging.Count > 0)
        {
            m[Dict.LOGGING] = new JsonArray();
            foreach( var log in logging)
            {
                m[Dict.LOGGING]?.AsArray().Add(log);
            }
        }

        m[Dict.DATE_TIME] = DateTime.UtcNow.ToString("o");

        if (verification != null && verification.Failed())
        {
            m[Dict.ERRORS] = verification.ToJson();
        }

        data[Dict.MEASUREMENTS]?.AsArray().Add(m);
    }

    /// <summary>
    /// Appends a measurement and the results to storage.
    /// </summary>
    /// <param name="measurement">The measurement to append.</param>
    /// /// <param name="results">The results to append.</param>
    /// <param name="comment">An optional comment for the measurement.</param>
    /// <param name="logging">Optional logging information.</param>
    /// <param name="verification">Optional verification information.</param>
    /// <remarks>
    /// Adds <c>date_time</c> in ISO-8601 UTC and optional <c>logging</c>/<c>errors</c> sections if provided.
    /// </remarks>
    public void AppendWithResults(Measurement measurement, Results results, string comment = "", List<string>? logging = null, Verification ? verification = null)
    {
        if (measurement == null)
            throw new ArgumentException("No measurement object provided to append!");

        var m = measurement.ToJson();

        if (results != null)
            m[Dict.RESULTS] = results.ToJson();

        if (!string.IsNullOrEmpty(comment))
        {
            m[Dict.COMMENT] = comment;
        }

        if (logging != null && logging.Count > 0)
        {
            m[Dict.LOGGING] = new JsonArray();
            foreach (var log in logging)
            {
                m[Dict.LOGGING]?.AsArray().Add(log);
            }
        }

        m[Dict.DATE_TIME] = DateTime.UtcNow.ToString("o");

        if(verification != null && verification.Failed())
        {
            m[Dict.ERRORS] = verification.ToJson();
        }

        data[Dict.MEASUREMENTS]?.AsArray().Add(m);
    }

    /// <summary>
    /// Saves the measurement data to a specified file.
    /// </summary>
    /// <param name="filename">The filename to save data to.</param>
    public void Save(string filename)
    {
        var options = new JsonSerializerOptions
        {
            NumberHandling = JsonNumberHandling.AllowReadingFromString | JsonNumberHandling.AllowNamedFloatingPointLiterals,
            WriteIndented = true,
            TypeInfoResolver = new DefaultJsonTypeInfoResolver()
        };

        using (var writer = new StreamWriter(filename))
        {
            var json = data.ToJsonString(options);
            writer.Write(json);
        }
    }

    /// <summary>
    /// Exports the measurement data as CSV next to the JSON file.
    /// </summary>
    /// <param name="filename">The JSON filename used to derive the CSV filename.</param>
    public void ExportAsCsv(string filename)
    {
        string csvFilename = Path.ChangeExtension(filename, ".csv");
        using var writer = new StreamWriter(csvFilename);

        writer.WriteLine("comment;air dark;air value;air ledPower;sample dark;sample value;sample ledPower;concentration;rfu");

        foreach (var node in data[Dict.MEASUREMENTS]?.AsArray() ?? new JsonArray())
        {
            if (node == null)
            {
                continue;
            }

            string comment = node[Dict.COMMENT]?.ToString() ?? string.Empty;
            string airDark = node[Dict.AIR]?[Dict.DARK]?.GetValue<double>().ToString(CultureInfo.InvariantCulture) ?? string.Empty;
            string airValue = node[Dict.AIR]?[Dict.VALUE]?.GetValue<double>().ToString(CultureInfo.InvariantCulture) ?? string.Empty;
            string airLedPower = node[Dict.AIR]?[Dict.LED_POWER]?.ToString() ?? string.Empty;
            string sampleDark = node[Dict.SAMPLE]?[Dict.DARK]?.GetValue<double>().ToString(CultureInfo.InvariantCulture) ?? string.Empty;
            string sampleValue = node[Dict.SAMPLE]?[Dict.VALUE]?.GetValue<double>().ToString(CultureInfo.InvariantCulture) ?? string.Empty;
            string sampleLedPower = node[Dict.SAMPLE]?[Dict.LED_POWER]?.ToString() ?? string.Empty;
            string concentration = node[Dict.RESULTS]?[Dict.CONCENTRATION]?.GetValue<double>().ToString(CultureInfo.InvariantCulture) ?? string.Empty;
            string rfu = node[Dict.RESULTS]?[Dict.RFU]?.GetValue<double>().ToString(CultureInfo.InvariantCulture) ?? string.Empty;

            writer.WriteLine(string.Join(";", new[]
            {
                comment,
                airDark,
                airValue,
                airLedPower,
                sampleDark,
                sampleValue,
                sampleLedPower,
                concentration,
                rfu
            }));
        }
    }

    /// <summary>
    /// Retrieves the list of stored measurements.
    /// </summary>
    /// <returns>A list of <see cref="Measurement"/> objects.</returns>
    public List<Measurement> Measurements()
    {
        var ret = new List<Measurement>();

        foreach (var m in data[Dict.MEASUREMENTS]?.AsArray() ?? new JsonArray())
            if (m != null)
                ret.Add(Measurement.FromJson(m));

        return ret;
    }

    /// <summary>
    /// Retrieves the list of stored results.
    /// </summary>
    /// <returns>A list of <see cref="Measurement"/> objects.</returns>
    public List<Results> Results()
    {
        var ret = new List<Results>();

        foreach (var m in data[Dict.MEASUREMENTS]?.AsArray() ?? new JsonArray())
        {
            if (m != null && m.AsObject().ContainsKey(Dict.RESULTS) && m[Dict.RESULTS] != null)
                ret.Add(EviFluor.Results.FromJson(m[Dict.RESULTS]));
        }

        return ret;
    }

    /// <summary>
    /// Gets the <see cref="StorageMeasurementEntry"/> at the specified index.
    /// </summary>
    /// <param name="index">The zero-based index of the measurement entry to retrieve.</param>
    /// <returns>
    /// The <see cref="StorageMeasurementEntry"/> located at the given index in the measurement list.
    /// </returns>
    /// <exception cref="IndexOutOfRangeException">
    /// Thrown if <paramref name="index"/> is outside the range of stored measurements.
    /// </exception>
    public StorageMeasurementEntry this[int index]
    {
        get
        {
            var measurementsArray = data[Dict.MEASUREMENTS]?.AsArray();
            if (measurementsArray == null || index < 0 || index >= measurementsArray.Count)
                throw new IndexOutOfRangeException("CustomRange index out of range");

            var node = measurementsArray.ElementAt(index);

            if(node == null)
                throw new IndexOutOfRangeException("CustomRange index out of range");

            return StorageMeasurementEntry.FromJson(node);
        }
    }

    /// <summary>
    /// Gets the number of stored measurements.
    /// </summary>
    public int Count => data[Dict.MEASUREMENTS]?.AsArray().Count ?? 0;
}

File: python/hse/evifluor/storage.py
# SPDX-License-Identifier: MIT
# SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

import csv
from datetime import datetime
import json
import os

from hse.evifluor import version
from hse.evifluor.constants import DictKeys
from hse.evifluor.measurement import Measurement, Results
from hse.evifluor.verification import Verification
from . import kits


class StorageMeasurementEntry:
    """Represents a stored measurement with optional comment, results, and raw JSON node."""

    def __init__(self, measurement, comment = None, results = None, node = None):
        """Initializes a storage entry with measurement data and optional metadata."""
        self.measurement = measurement
        self.comment     = comment
        self.results     = results
        self.node        = node
        
    def __repr__(self):
        """Returns a textual representation of the storage entry."""
        return "measurement:{} comment:{} results:{}".format(self.measurement, self.comment, self.results)
        
    def has_results(self):
        """Indicates whether calculated results are present."""
        if self.results is None:
            return False
        else:
            return True
            
    def apply_results(self, factors, kit = kits.Default()):
        """Applies calibration factors to compute results and updates the JSON node."""
        results = self.measurement.results(factors, kit)
        self.node[DictKeys.RESULTS] = results.to_json()
        if DictKeys.ERRORS in self.node:
            v = Verification.from_json(self.node[DictKeys.ERRORS])
        else:
            v = Verification()
        v.check(results)
        if v.failed():
            self.node[DictKeys.ERRORS] = v.to_json()
        
    @staticmethod
    def from_json(node):
        """Creates a StorageMeasurementEntry from a JSON node."""
        if DictKeys.RESULTS in node:
            results = Results.from_json(node[DictKeys.RESULTS])
        else:
            results = None

        if DictKeys.COMMENT in node:
            comment = node[DictKeys.COMMENT]
        else:
            comment = None
            
        return StorageMeasurementEntry(Measurement.from_json(node), comment, results, node)
        

class StorageMeasurement:
    """Handles storage and retrieval of measurements using JSON serialization."""

    def __init__(self, filename = None):
        """Initializes storage from an optional filename or creates an empty structure."""
        
        if filename is None:
            self.data = {DictKeys.MEASUREMENTS: []}
        else:
            with open(filename, 'rb') as f:
                self.data = json.load(f)
    
    def append(self, measurement, comment = None, logging = None, verification = None):
        """Appends a measurement and optional metadata to storage."""
        if isinstance(measurement, Measurement):
            m = measurement.to_json()
        else:
            raise Exception("Class not supported!")
        
        if comment:
            m[DictKeys.COMMENT] = comment

        if logging:
            m[DictKeys.LOGGING] = logging
            
        m[DictKeys.DATE_TIME] = datetime.now().isoformat()
            
        if verification != None and verification.failed():
            m[DictKeys.ERRORS] = verification.to_json()
            
        self.data[DictKeys.MEASUREMENTS].append(m)

    def append_with_results(self, measurement, results, comment = None, logging = None, verification = None):
        """Appends a measurement along with computed results and metadata."""
        if isinstance(measurement, Measurement):
            m = measurement.to_json()
        else:
            raise Exception("Class not supported!")
        
        if results:
            m[DictKeys.RESULTS] = results.to_json()
            
        if comment:
            m[DictKeys.COMMENT] = comment
            
        if logging:
            m[DictKeys.LOGGING] = logging
            
        m[DictKeys.DATE_TIME] = datetime.now().isoformat()

        if verification != None and verification.failed():
            m[DictKeys.ERRORS] = verification.to_json()
            
        self.data[DictKeys.MEASUREMENTS].append(m)

        
    def add_device_info(self, device, comment = None):
        """Adds device information and metadata to the storage document."""
        self.data[DictKeys.INFO] = { "date"              : datetime.now().isoformat(),
                                     "product"           : "eviFluor",
                                     "production_number" : device.production_number(),
                                     "serial_number"     : device.serial_number(),
                                     "firmware_version"  : device.firmware_version(),
                                     "comment"           : comment or "",
                                     "api"               : version.VERSION
                               }
                               
    def add_dict(self, key, value):
        """Inserts an arbitrary key/value pair into the storage document."""
        self.data[key] = value

    def save(self, filename):
        """Saves the storage data to disk in JSON format."""
        with open(filename, 'w') as f:
            json.dump(self.data, f, indent=2)
            
    def measurements(self):
        """Returns the measurements reconstructed from storage."""
        ret = []
        for m in self.data[DictKeys.MEASUREMENTS]:
            ret.append(Measurement.from_json(m))
        return ret

    def results(self):
        """Returns the stored results."""
        ret = []
        for m in self.data[DictKeys.MEASUREMENTS]:
            ret.append(Results.from_json(m[DictKeys.RESULTS]))
        return ret
   
    def __getitem__(self, item):
        """Retrieves the entry at the specified index."""
        if item >= self.__len__():
            raise IndexError("CustomRange index out of range")
        return StorageMeasurementEntry.from_json(self.data[DictKeys.MEASUREMENTS][item])

    def __len__(self):
        """Returns the number of stored measurements."""
        return len(self.data[DictKeys.MEASUREMENTS])
    
    @staticmethod
    def export_as_csv(filenameJson, flat = False):
        """Exports stored measurements to CSV, optionally flattening channel values."""
        data = {}
        if os.path.isfile(filenameJson):
            with open(filenameJson, 'rb') as f:
                data = json.load(f)
        else:
            raise Exception("File {} not found".format(filenameJson))

        filenameCsv, _ = os.path.splitext(filenameJson)
        with open('{}.csv'.format(filenameCsv), 'w', newline='') as csvfile:
            if flat == True:
                fieldnames = [
                    DictKeys.COMMENT,
                    DictKeys.AIR_DARK,
                    DictKeys.AIR_VALUE,
                    DictKeys.AIR_LED_POWER,
                ]
                writer = csv.writer(csvfile, delimiter=";")
                for measurement in data[DictKeys.MEASUREMENTS]:
                    row = []
                    row.append(measurement.get(DictKeys.COMMENT, ""))
                    for value in measurement[DictKeys.VALUES]:
                        row.append(value[DictKeys.DARK])
                        row.append(value[DictKeys.VALUE])
                        row.append(value[DictKeys.LED_POWER])
                    print(row)
                    writer.writerow(row)
            else:
                fieldnames = [
                    DictKeys.COMMENT,
                    DictKeys.AIR_DARK,
                    DictKeys.AIR_VALUE,
                    DictKeys.AIR_LED_POWER,
                    DictKeys.SAMPLE_DARK,
                    DictKeys.SAMPLE_VALUE,
                    DictKeys.SAMPLE_LED_POWER,
                    DictKeys.CONCENTRATION,
                    DictKeys.RFU,
                ]

                writer = csv.DictWriter(csvfile, fieldnames=fieldnames, delimiter=";")
                writer.writeheader()
                for measurement in data[DictKeys.MEASUREMENTS]:
                    if measurement.get(DictKeys.SAMPLE) != None:
                        if DictKeys.RESULTS in measurement:
                            concentration = measurement[DictKeys.RESULTS][DictKeys.CONCENTRATION]
                            rfu = measurement[DictKeys.RESULTS].get(DictKeys.RFU, "")
                        else:
                            concentration = ""
                            rfu = ""

                        if DictKeys.AIR in measurement:
                            air_dark      = measurement[DictKeys.AIR][DictKeys.DARK]
                            air_value     = measurement[DictKeys.AIR][DictKeys.VALUE]
                            air_led_power = measurement[DictKeys.AIR][DictKeys.LED_POWER]
                        else:
                            air_dark      = ""
                            air_value     = ""
                            air_led_power = ""

                        if DictKeys.COMMENT in measurement:
                            comment       = measurement[DictKeys.COMMENT]
                        else:
                            comment       = ""

                        if DictKeys.SAMPLE in measurement:
                            sample_dark      = measurement[DictKeys.SAMPLE][DictKeys.DARK]
                            sample_value     = measurement[DictKeys.SAMPLE][DictKeys.VALUE]
                            sample_led_power = measurement[DictKeys.SAMPLE][DictKeys.LED_POWER]
                        else:
                            sample_dark      = ""
                            sample_value     = ""
                            sample_led_power = ""

                        if DictKeys.COMMENT in measurement:
                            comment       = measurement[DictKeys.COMMENT]
                        else:
                            comment       = ""

                        writer.writerow({
                            DictKeys.COMMENT:         comment,
                            DictKeys.AIR_DARK:        air_dark,
                            DictKeys.AIR_VALUE:       air_value,
                            DictKeys.AIR_LED_POWER:   air_led_power,
                            DictKeys.SAMPLE_DARK:     sample_dark,
                            DictKeys.SAMPLE_VALUE:    sample_value,
                            DictKeys.SAMPLE_LED_POWER: sample_led_power,
                            DictKeys.CONCENTRATION:   concentration,
                            DictKeys.RFU:             rfu,
                        })

    @staticmethod
    def exportAsCsv(filenameJson, flat = False):
        """Backwards-compatible alias for export_as_csv."""
        return __class__.export_as_csv(filenameJson, flat)
